### PR TITLE
[subset] expose subset plan in public subsetting API

### PR DESCRIPTION
**Pull request by garretrieger**
For #3140 

Provides access to the subset plan so callers can find information on things such as the glyph mapping from the subset.

### CODE REVIEW

**Comment by behdad** [comment]
To me not reusing a hb-map in the API would be a mistake. We should just return a hb-map and let that API be reused.

**Comment by garretrieger** [comment]
> To me not reusing a hb-map in the API would be a mistake. We should just return a hb-map and let that API be reused.

Sounds good. I'll update this.

File: src/hb-subset.h [hunk 1]
```diff
@@ -145,6 +154,28 @@ hb_subset_input_set_flags (hb_subset_input_t *input,
 HB_EXTERN hb_face_t *
 hb_subset_or_fail (hb_face_t *source, const hb_subset_input_t *input);
 
+HB_EXTERN hb_face_t *
+hb_subset_from_plan_or_fail (hb_subset_plan_t *plan);
```
Comment (behdad) [inline src/hb-subset.h:158]: Borrowing name from shape_plan would suggest:
```
hb_subset_plan_execute_or_fail
```

File: src/hb-subset.h [hunk 2]
```diff
@@ -145,6 +154,28 @@ hb_subset_input_set_flags (hb_subset_input_t *input,
...
+
+HB_EXTERN hb_subset_plan_t *
+hb_subset_plan_create (hb_face_t           *face,
+		       const hb_subset_input_t   *input);
+
+HB_EXTERN void
+hb_subset_plan_destroy (hb_subset_plan_t *plan);
```
Comment (behdad) [inline src/hb-subset.h:165]: Please `_reference`, `_set_user_data`, `_get_user_data`.

File: src/hb-subset.h [hunk 3]
```diff
@@ -145,6 +154,25 @@ hb_subset_input_set_flags (hb_subset_input_t *input,
 HB_EXTERN hb_face_t *
 hb_subset_or_fail (hb_face_t *source, const hb_subset_input_t *input);
 
+HB_EXTERN hb_face_t *
+hb_subset_from_plan_or_fail (hb_subset_plan_t *plan);
+
+HB_EXTERN hb_subset_plan_t *
+hb_subset_plan_create (hb_face_t           *face,
```
Comment (behdad) [inline src/hb-subset.h:161]: Probably also `_create_or_fail` instead of `_create`.

Comment (garretrieger) [reply src/hb-subset.h:161]: plan_create always returns a valid pointer. It uses an in_error() to check for failure.

Comment (behdad) [reply src/hb-subset.h]: Would it be useful for our purposes though?

Comment (garretrieger) [reply src/hb-subset.h]: Yeah I could see that (nullptr) being more intuitive for end users. I'll change create to be or_fail.

File: src/hb-subset.cc [hunk 4]
```diff
@@ -343,17 +343,38 @@ hb_subset_or_fail (hb_face_t *source, const hb_subset_input_t *input)
...
+  }
+
+  hb_face_t * result = hb_subset_plan_execute_or_fail (plan);
+  hb_subset_plan_destroy (plan);
+  return result;
+}
+
+
+/**
+ * hb_subset_from_plan_or_fail:
```
Comment (khaledhosny) [inline src/hb-subset.cc:358]: This needs to match the actual function name.

File: src/hb-subset.cc [hunk 5]
```diff
@@ -343,17 +343,38 @@ hb_subset_or_fail (hb_face_t *source, const hb_subset_input_t *input)
...
+ * @plan: a subsetting plan.
+ *
+ * Subsets a font according to provided plan. Returns nullptr
+ * if the subset operation fails.
+ *
```
Comment (khaledhosny) [inline src/hb-subset.cc:363]: The return value needs to be documented as well.

File: src/hb-subset-plan.cc [hunk 6]
```diff
@@ -596,3 +602,107 @@ hb_subset_plan_destroy (hb_subset_plan_t *plan)
...
 }
+
+/**
+ * hb_subset_plan_old_to_new_glyph_mapping:
+ * @plan: a subsetting plan.
+ *
+ * Returns the mapping between glyphs in the original font to glyphs
+ * in the subset that will be produced by @plan.
+ *
+ * Since: REPLACEME
```
Comment (khaledhosny) [inline src/hb-subset-plan.cc:613]: The return value needs to be documented (with `Return value: ` syntax)

**Comment by behdad** [comment]
Lgtm

File: src/hb-subset-plan.cc [hunk 7]
```diff
@@ -596,3 +602,108 @@ hb_subset_plan_destroy (hb_subset_plan_t *plan)
...
+ * hb_subset_plan_old_to_new_glyph_mapping:
+ * @plan: a subsetting plan.
+ *
+ * Returns the mapping between codepoints in the original font and the
+ * associated glyph id in the original font.
+ *
+ * Since: REPLACEME
+ **/
+const hb_map_t*
+hb_subset_plan_codepoint_to_old_glyph_mapping (const hb_subset_plan_t *plan)
```
Comment (behdad) [inline src/hb-subset-plan.cc:647]: Should we even return this? This is available from the original face readily, right?

Comment (garretrieger) [reply src/hb-subset-plan.cc:647]: In the case where the subset was specified using only glyph id's this provides a method to determine which codepoints are going to be in the final subset.

Comment (behdad) [reply src/hb-subset-plan.cc:647]: I see. Sure.

**Pull request merged**